swaybg/main.c

```diff
@@ -190,24 +202,90 @@ static void xdg_output_handle_logical_size(void *data,
 	// Who cares
 }
 
+static void find_config(struct swaybg_output *output, const char *name) {
+	struct swaybg_output_config *config = NULL;
+	wl_list_for_each(config, &output->state->configs, link) {
+		if (strcmp(config->output, name) == 0) {
+			output->config = config;
+			return;
+		} else if (!output->config && strcmp(config->output, "*") == 0) {
+			output->config = config;
+		}
+	}
+}
+
 static void xdg_output_handle_name(void *data,
 		struct zxdg_output_v1 *xdg_output, const char *name) {
 	struct swaybg_output *output = data;
-	struct swaybg_state *state = output->state;
-	if (strcmp(name, state->args->output) == 0) {
-		assert(state->output == NULL);
-		state->output = output;
+	output->name = strdup(name);
+
+	// If description was sent first, the config may already be populated. If
+	// there is an identifier config set, keep it.
+	if (!output->config || strcmp(output->config->output, "*") == 0) {
+		find_config(output, name);
 	}
 }
 
 static void xdg_output_handle_description(void *data,
 		struct zxdg_output_v1 *xdg_output, const char *description) {
-	// Who cares
+	struct swaybg_output *output = data;
+
+	// wlroots currently sets the description to `make model serial (name)`
+	// If this changes in the future, this will need to be modified.
+	char *paren = strrchr(description, '(');
+	if (paren) {
+		size_t length = paren - description;
+		output->identifier = malloc(length);
+		if (!output->identifier) {
+			sway_log(SWAY_ERROR, "Failed to allocate output identifier");
+			return;
+		}
+		strncpy(output->identifier, description, length);
+		output->identifier[length - 1] = '\0';
+
+		find_config(output, output->identifier);
+	}
+}
+
+static void create_layer_surface(struct swaybg_output *output) {
+	output->surface = wl_compositor_create_surface(output->state->compositor);
+	assert(output->surface);
+
+	// Empty input region
+	output->input_region =
```

Comment:
I don't think we need to save this. We can destroy it right away.

Reply:
Doesn't it need to be stored so the `wl_region` can be destroyed later?

Reply:
We can destroy it right away after `wl_surface_set_input_region`.

Reply:
Changed